Tests/BaseLib/TestFunctional.cpp

```diff
@@ -0,0 +1,239 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <gtest/gtest.h>
+#include <logog/include/logog.hpp>
+
+#include "BaseLib/Functional.h"
+
+class InstanceCounter
+{
+public:
+    InstanceCounter() {
+        ++_num_constructed;
+    }
+    InstanceCounter(InstanceCounter const&) {
+        ++_num_copied;
+    }
+    InstanceCounter(InstanceCounter&&) {
+        ++_num_moved;
+    }
+    ~InstanceCounter() {
+        ++_num_destroyed;
+    }
+
+    static int getNumberOfConstructions() { return _num_constructed; }
+    static int getNumberOfCopies() { return _num_copied; }
+    static int getNumberOfMoves() { return _num_moved; }
+    static int getNumberOfDestructions() { return _num_destroyed; }
+    static int getNumberOfInstances()
+    {
+        return _num_constructed + _num_moved + _num_copied - _num_destroyed;
+    }
+
+private:
+    static int _num_constructed;
+    static int _num_copied;
+    static int _num_moved;
+    static int _num_destroyed;
+};
+
+int InstanceCounter::_num_constructed = 0;
+int InstanceCounter::_num_copied = 0;
+int InstanceCounter::_num_moved = 0;
+int InstanceCounter::_num_destroyed = 0;
+
+class A : public InstanceCounter
+{
+public:
+    A(const double value) : _value(value) {}
+    A(A const& other) : InstanceCounter(other), _value(other._value) {}
+    A(A&& other) : InstanceCounter(std::move(other)), _value(other._value) {}
+    A& operator=(A const& other) { _value = other._value; return *this; }
+    A& operator=(A&& other) { _value = other._value; return *this; }
+
+    void add(A const& other) { _value += other._value; }
+    // pass by value intended.
+    void multiply(A other) {
+        _value *= other._value;
+        other._value = 0.0;
+    }
+    double getValue() const { return _value; }
+    double& getValueRef() { return _value; }
+    double operator()(double const x) { return _value*x; }
+private:
+    double _value;
+};
+
+#define EXPECT_INSTANCES(num_const, num_move, num_copy, num_inst) \
+    EXPECT_EQ((num_const), InstanceCounter::getNumberOfConstructions());    \
+    EXPECT_EQ((num_move), InstanceCounter::getNumberOfMoves());             \
+    EXPECT_EQ((num_copy), InstanceCounter::getNumberOfCopies());            \
+    EXPECT_EQ((num_inst), InstanceCounter::getNumberOfInstances())
+
+#define UPDATE_INSTANCES(num_const, num_move, num_copy, num_inst) \
+    num_const = InstanceCounter::getNumberOfConstructions();              \
+    num_move = InstanceCounter::getNumberOfMoves();                       \
+    num_copy = InstanceCounter::getNumberOfCopies();                      \
+    num_inst = InstanceCounter::getNumberOfInstances()
```

Comment:
the above must be a macro in order to get the right line in gtest's messages. I'll change the method below.